lib/component.js

```diff
@@ -805,6 +819,114 @@ class Component extends WebComponent {
+
+    // vuejs like this.data.$data to reference "real" value of params
+    Object.defineProperty(this.params, `$params`, {
+      enumerable: false,
+      value: {},
+    });
+
+    for (const param of Object.keys(paramTypes)) {
+      // convert type shorthand to object
+      let paramType = paramTypes[param];
+      if (PARAM_TYPE_DEFAULTS.has(paramType)) {
+        paramType = {type: paramType};
+      }
+
+      // Ensure param type is valid
+      const type = paramType.type;
+      if (!PARAM_TYPE_DEFAULTS.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for attr: ${param} in paramType. ` +
+            `Only (${Array.from(PARAM_TYPE_DEFAULTS.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      if (paramType.default && paramType.required) {
+        throw new Error(`${this}: param '${param}' cannot have both required and default`);
+      }
+
+      if ([Object, Array, Map, Set].includes(type) && !paramType.shouldUpdate) {
+        throw new Error(
+          `Object like params '${param}' on ${this.constructor.name} must have 'shouldUpdate' function passed in`,
+        );
+      }
+
+      const paramSchemaObj = {
+        type,
+        default: paramType.default ? paramType.default : PARAM_TYPE_DEFAULTS.get(type),
+        required: paramType.required ? paramType.required : false,
+        // by default use === for comparision
+        shouldUpdate: paramType.shouldUpdate || ((oldVal, newVal) => oldVal !== newVal),
+      };
+
+      // set default value for the params
+      this.params.$params[param] = paramSchemaObj.default;
+
+      this._paramTypes[param] = paramSchemaObj;
+    }
+
+    const paramDescription = Object.keys(paramTypes)
+      .map((param) => [
+        param,
+        {
+          get() {
+            return this.$params[param];
+          },
+        },
+      ])
+      .reduce((acc, cur) => {
+        acc[cur[0]] = cur[1];
+        return acc;
+      }, {});
+    Object.defineProperties(this.params, paramDescription);
+    Object.freeze(this.params);
+    return this._paramTypes;
+  }
+
+  _enqueueParamUpdate(param, value) {
+    this._count = 0;
+    if (this._queuedParamUpdates) {
+      this._queuedParamUpdates[param] = value;
+      return;
+    }
+    this._queuedParamUpdates = {[param]: value};
+    queueMicrotask(() => {
+      const options = {
+        cascade: false,
+        store: `params`,
+      };
+      this.runHook(`preUpdate`, options, this.state, Object.assign({}, this.params.$params, this._queuedParamUpdates));
+      Object.assign(this.params.$params, this._queuedParamUpdates);
+      this.domPatcher.update(this.state);
+      this.runHook(`postUpdate`, options, this.state, this.params);
+      this._queuedParamUpdates = null;
+    });
+  }
+
+  _setParams(params) {
```

Comment:
Is there a particular reason why you've made the `_setParams()` method effectively private but prefixing it with an underscore `_`? I think it's a good idea to make it a public method so that non-JSX code can imperatively use it as well.

Reply:
hmm i assume that except for panel, no one should use this method to set params, even want to use a symbo further hide it. what's the usecase for exposing this method outside?

Reply:
Eg you have a test case:
```typescript
it(`renders line series`, function () {
  const myChart = new InsightsLineChart();
  document.body.appendChild(myChart);
  expect(myChart.state.series).to.equal([]);
  myChart.setParams({bookmark: INSIGHTS_BOOKMARK, results: INSIGHTS_QUERY_RESULTS});
  expect(myChart.state.series).to.be.of.length(6);
});
```

Sometimes we may want to manually create & attach elements in the DOM, eg
```typescript
downloadInvoicePDF(url) {
  const link = document.createElement(`mp-download-link`);

  // launch named download via hidden link
  link.setParams({href: url, download: `invoice.pdf`});
  link.click();
}
```

Reply:
hmm ok that makes sense, though the downloadPDF thing seems an anti pattern for me 🙈 

for the testcases, i think it should be sth like this (passing in the para with `init`)

```
it(`renders line series`, function () { 
  const comp = init(`insight-chart`, {
    params: {
      bookmark: INSIGHTS_BOOKMARK,
      results: INSIGHTS_QUERY_RESULTS,
    }
  });
  const myChart = mount(comp)
  expect(myChart.state.series).to.equal([]); 
  expect(myChart.state.series).to.be.of.length(6); 
});
```

Reply:
Sure but `init()` itself is going to need to call `component.setParams()` to apply params because params is an artificial Panel concept.

Also, it seems likely that tests will want to change the params during a test case. Your own code example simultaneously expects `myChart.state.series` to be equal to an empty array `[]` and be `of.length(6)`. Those two expect statements were supposed to separately check conditions before & after a params change

Reply:
right just make `_setParams` to `setParams` to indicate it's not strictly private method

Reply:
Yeah. Generally nothing outside Panel's own code should be using methods prefixed with underscores because the prefix conventionally means Panel can change or break compatibility with these private methods at any time.